enhancements/cluster-api/crd-compatibility-checker.md

```diff
@@ -0,0 +1,469 @@
+Assuming that the CRD from version Y is compatible with the one for version X required by CCAPIO, the CRD is admitted and everything works as normal.
+
+#### CCAPIO prevents cluster upgrade if it would conflict with an existing CRD
+
+This example covers the scenario when CCAPIO is first enabled in OCP, but more generally if CCAPIO starts using a CRD which HCP is currently using.
+We want to prevent upgrade if a CRD currently in use would be incompatible with the version required by CCAPIO after upgrade.
+
+OCP 4.Y is the current version.
+OCP 4.Y+1 is the upgraded version.
+
+Some time during the OCP 4.Y lifecycle we add an additional set of transport config maps to 4.Y.z which contain the CRDs required by 4.Y+1.
+CCAPIO scans these, but does not load them.
+For each CRD it creates a new CRDCompatibilityRequirement.
+e.g. for the `Cluster` CRD it would create:
+
+```yaml
+apiVersion: operator.openshift.io/v1alpha1
+kind: CRDCompatibilityRequirement
+metadata:
+  name: cluster-api-cluster-ccapio-4.Y+1
+spec:
+  crdRef: clusters.cluster.x-k8s.io
+  creatorDescription: "OpenShift Cluster CAPI Operator for OCP 4.Y+1"
+  compatibilityCRD: |
+    ...
+    <complete YAML document of future Cluster CRD from transport config map>
+    ...
+  crdAdmitAction: Warn
+```
+
+The differences from the previous example are:
+* The requirement identifies itself as being related to a different version.
+* `crdAdmitAction` is `Warn` instead of `Enforce`, so this requirement will not prevent updates to the current CRD.
+* There is no `objectSchemaValidation`.
+
+CCAPIO will mark itself as not upgradeable if any CRDCompatibilityRequirement created for an upgrade transport config map reports its `Compatible` condition as `False`.
+
+### API Extensions
+
+The following is an outline of the CRDCompatibilityRequirement CRD:
+
+```yaml
+apiVersion: operator.openshift.io/v1alpha1
+kind: CRDCompatibilityRequirement
+metadata:
+  ...
+spec:
+  # Fields to be implemented initially
+  crdRef: <the name of the target CRD>
+  creatorDescription: |
+    <human readable description of who created this requirement. displayed in errors and warnings the validation webhooks>
+  compatibilityCRD: |
+    <the complete CRD required by the creator of this requirement, as yaml>
+  crdAdmitAction: <Enforce or Warn>
+
+  # If defined, also validate that matching objects conform to the schema given in compatibilityCRD.
+  # This will be implemented in a second phase. The initial API will not contain this field.
+  objectSchemaValidation:
+    # namespaceSelector, objectSelector, and matchConditions will be copied to
+    # the corresponding ValidatingWebhookConfiguration. Their definitions and
+    # semantics are therefore identical to those in
+    # ValidatingWebhookConfiguration.
+    namespaceSelector:
+      matchLabels:
+        "kubernetes.io/metadata.name": "openshift-cluster-api"
+    objectSelector:
+      ...
+    matchConditions:
+      ...
+    action: <Enforce or Warn>
```

Comment:
Again, what is enforce and warn actually doing here? Is the same as above? Perhaps some blurb below this would help explain what they mean